File: lib/ios/bridge/WEPersonalizationBridge.swift
import UIKit
import React
import WEPersonalization
import WebEngage

#if RCT_NEW_ARCH_ENABLED
import WEPersonalizationSpec
#endif

@objc(WEPersonalizationBridge)
class WEPersonalizationBridge: RCTEventEmitter {
    
#if RCT_NEW_ARCH_ENABLED
    // MARK: - TurboModule conformance
    static func moduleName() -> String! {
        return "WEPersonalizationBridge"
    }
    
    static func requiresMainQueueSetup() -> Bool {
        return false
    }
#endif
    
    public static var emitter: RCTEventEmitter!
    var propertyId = 0;
    var weCampaignData: WECampaignData? = nil
    var customPropertiesList = [Int]()
    var WEGPEPluginVersion = "1.1.1"
    
    
    static let shared = WEPersonalizationBridge()
    override init() {
        super.init()
        WELogger.initLogger()
        initialiseWEGVersion()
        WEPersonalizationBridge.emitter = self
        WEPersonalization.shared.initialise()
        UserDefaults.standard.setValue(false, forKey: WEPersonalization.Constants.KEY_SHOULD_AUTO_TRACK_IMPRESSIONS)
        WEPersonalization.shared.registerPropertyRegistryCallbacks(WECampaignCallbackHandler.shared)
        
    }
    
    func initialiseWEGVersion() {
        let key: WegVersionKey = .RNPE
        WebEngage.sharedInstance().setVersionForChildSDK(WEGPEPluginVersion, for: key)
      }
    
    @objc func initWePersonalization() {
        
    }
    
    @objc func registerWECampaignCallback() {
        WELogger.d(WEConstants.TAG+"WEP: WEPersonalizationBridge: registerWECampaignCallback called ")
        WEPersonalization.shared.registerWECampaignCallback(WECampaignCallbackHandler.shared)
    }
    
    @objc func deregisterWECampaignCallback() -> Void {
        WELogger.d(WEConstants.TAG+" WEP: WEPersonalizationBridge: unRegisterForCampaigns called")
        WEPersonalization.shared.unregisterWECampaignCallback(WECampaignCallbackHandler.shared)
    }
    
    @objc func registerProperty(_ propertyId: String, screenName: String) {
        WELogger.d(WEConstants.TAG+" WEP: WEPersonalizationBridge: registerProperty called - \(propertyId)")
        if let intPropertyId = Int(propertyId) {
            self.propertyId = intPropertyId
            let data: [String: Any] = [
                WEConstants.PAYLOAD_ID: intPropertyId,
                WEConstants.PAYLOAD_SCREEN_NAME: screenName,
                WEConstants.PAYLOAD_IOS_PROPERTY_ID: intPropertyId
            ]
            WECustomPropertyRegistry.instance.registerData(map: data)
            WEPersonalization.shared.registerWEPlaceholderCallback(intPropertyId, CustomCallbackHandler.shared.self)
        }
    }
    
    
    @objc func deregisterProperty(_ propertyId: String) {
        WELogger.d(WEConstants.TAG+" WEP: WEPersonalizationBridge: deregisterProperty called - \(propertyId)")
        if let intPropertyId = Int(propertyId) {
            WECustomPropertyRegistry.instance.removeRegisterData(id: intPropertyId)
            WEPersonalization.shared.unregisterWEPlaceholderCallback(intPropertyId)
        }
    }
    
    @objc func trackClick(_ propertyId: String, attributes: [String: Any]) -> Void {
        if let intPropertyId = Int(propertyId) {
            let weginline = WECustomPropertyRegistry.instance.getWEGHinline(targetViewTag: intPropertyId)
            weginline?.campaignData?.trackClick(attributes: attributes)
        }
    }
    
    @objc func trackImpression(_ propertyId: String, attributes: [String: Any]) -> Void {
        if let intPropertyId = Int(propertyId) {
            let weginline = WECustomPropertyRegistry.instance.getWEGHinline(targetViewTag: intPropertyId)
            weginline?.campaignData?.trackImpression(attributes: attributes)
        }
    }
    
    @objc override func addListener(_ eventType: String) {
        // Handled by React Native
    }
    
    @objc override func removeListeners(_ count: Double) {
        // Handled by React Native
    }
    
    
    
    open override func supportedEvents() -> [String] {
        ["onDataReceived", "onRendered", "onPropertyCacheCleared", "onPlaceholderException", "testAk", "onCampaignPrepared", "onCampaignClicked", "onCampaignException", "onCampaignShown", "onCustomDataReceived", "onCustomPlaceholderException"]
    }
}

File: lib/android/src/main/java/com/webengagepersonalization/bridge/WEPersonalizationBridge.java
package com.webengagepersonalization.bridge;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Arguments;
import com.webengage.personalization.WEPersonalization;
import com.webengage.personalization.callbacks.WECampaignCallback;
import com.webengage.personalization.callbacks.WEPlaceholderCallback;
import com.webengage.personalization.data.WECampaignData;
import com.webengage.sdk.android.Logger;
import com.webengagepersonalization.utils.WEUtils;
import com.webengagepersonalization.utils.WEConstants;
import com.webengagepersonalization.registry.WECustomPropertyRegistry;

import androidx.annotation.Nullable;

@ReactModule(name = WEConstants.PERSONALIZATION_BRIDGE)
public class WEPersonalizationBridge extends ReactContextBaseJavaModule implements WEPlaceholderCallback, WECampaignCallback {
  private ReactApplicationContext applicationContext = null;

  public WEPersonalizationBridge(ReactApplicationContext reactContext) {
    super(reactContext);
    this.applicationContext = reactContext;
    WEPersonalization.Companion.get().init();
  }

  @ReactMethod
  public void initWePersonalization() {}
  
  @ReactMethod
  public void registerProperty(String tagName, String screenName) {
    WEPersonalization.Companion.get().registerWEPlaceholderCallback(tagName, this);
  }

  @ReactMethod
  public void deregisterProperty(String tagName) {
    WEPersonalization.Companion.get().unregisterWEPlaceholderCallback(tagName);
  }


  @ReactMethod
  public void registerWECampaignCallback() {
    WEPersonalization.Companion.get().registerWECampaignCallback(this);
  }

  @ReactMethod
  public void deregisterWECampaignCallback() {
    WEPersonalization.Companion.get().unregisterWECampaignCallback(this);
  }

  @ReactMethod
  public void trackImpression(String propertyId, ReadableMap attributes) {
    WECampaignData weCampaignData = WECustomPropertyRegistry.get().getMapData(propertyId);
    if (weCampaignData != null) {
      weCampaignData.trackImpression(WEUtils.convertHybridMapToNativeMap(attributes));
    }
  }

  @ReactMethod
  public void trackClick(String propertyId, ReadableMap attributes) {
    WECampaignData weCampaignData = WECustomPropertyRegistry.get().getMapData(propertyId);
    if (weCampaignData != null) {
      weCampaignData.trackClick(WEUtils.convertHybridMapToNativeMap(attributes));
    }
  }

  @Override
  public boolean onCampaignClicked(@NonNull String actionId, @NonNull String deepLink, @NonNull WECampaignData weCampaignData) {
    Logger.d(WEConstants.TAG, "WEPersonalizationBridge: onCampaignClicked actionId- " + actionId + " \n deepLink- " + deepLink);
    WritableMap params = Arguments.createMap();
    params = WEUtils.generateParams(actionId, deepLink, weCampaignData);
    WEUtils.sendEventToHybrid(applicationContext, "onCampaignClicked", params);
    return true;
  }

  @Override
  public void onCampaignException(@Nullable String campaignId, @NonNull String targetViewId, @NonNull Exception e) {
    Logger.d(WEConstants.TAG, "WEPersonalizationBridge: onCampaignException " + targetViewId);
    WritableMap params = Arguments.createMap();
    params = WEUtils.generateParams(campaignId, targetViewId, e);
    WEUtils.sendEventToHybrid(applicationContext, "onCampaignException", params);
  }

  @Nullable
  @Override
  public WECampaignData onCampaignPrepared(@NonNull WECampaignData weCampaignData) {
    Logger.d(WEConstants.TAG, "WEPersonalizationBridge: onCampaignPrepared " + weCampaignData.getTargetViewId());
    WritableMap params = Arguments.createMap();
    params = WEUtils.generateParams(weCampaignData);
    WEUtils.sendEventToHybrid(applicationContext, "onCampaignPrepared", params);
    return null;
  }

  @Override
  public void onCampaignShown(@NonNull WECampaignData weCampaignData) {
    Logger.d(WEConstants.TAG, "WEPersonalizationBridge: onCampaignShown " + weCampaignData.getTargetViewId());
    WritableMap params = Arguments.createMap();
    params = WEUtils.generateParams(weCampaignData);
    WEUtils.sendEventToHybrid(applicationContext, "onCampaignShown", params);
  }

  @Override
  @NonNull
  public String getName() {
    return WEConstants.PERSONALIZATION_BRIDGE;
  }

  @Override
  public void onDataReceived(WECampaignData weCampaignData) {
    WritableMap params = Arguments.createMap();
    params = WEUtils.generateParams(weCampaignData);
    String targetView = weCampaignData.getTargetViewId();
    WECustomPropertyRegistry.get().registerProperty(targetView, weCampaignData);
    WEUtils.sendEventToHybrid(applicationContext, "onCustomDataReceived", params);
  }

  @Override
  public void onPlaceholderException(String campaignId, String targetViewId, Exception e) {
    WritableMap params = Arguments.createMap();
    params = WEUtils.generateParams(campaignId, targetViewId, e);
    WEUtils.sendEventToHybrid(applicationContext, "onCustomPlaceholderException", params);
  }

  @Override
  public void onRendered(WECampaignData weCampaignData) {
  }
}
